aptos-move/framework/aptos-framework/sources/nonce_validation.move

```diff
@@ -0,0 +1,207 @@
+                    last_stored_time: 0,
+                },
+            ]
+        };
+        bucket.nonce_maps[0].nonces.allocate_spare_slots(5);
+        bucket.nonce_maps[1].nonces.allocate_spare_slots(5);
+        bucket
+    }
+
+    // This method is used to prefill the nonce_table with empty buckets one by one.
+    public entry fun add_nonce_buckets(count: u64) acquires NonceHistory {
+        assert!(exists<NonceHistory>(@aptos_framework), error::invalid_state(E_NONCE_HISTORY_DOES_NOT_EXIST));
+        let nonce_history = &mut NonceHistory[@aptos_framework];
+        for (i in 0..count) {
+            if (nonce_history.next_key <= NUM_BUCKETS) {
+                if (!nonce_history.nonce_table.contains(nonce_history.next_key)) {
+                    nonce_history.nonce_table.add(
+                        nonce_history.next_key,
+                        empty_bucket()
+                    );
+                };
+                nonce_history.next_key = nonce_history.next_key + 1;
+            }
+        }
+    }
+
+    // Returns true if the input (address, nonce) pair doesn't exist in the nonce history, and inserted into nonce history successfully.
+    // Returns false if the input (address, nonce) pair already exists in the nonce history.
+    public(friend) fun check_and_insert_nonce(
+        sender_address: address,
+        nonce: u64,
+        txn_expiration_time: u64,
+    ): bool acquires NonceHistory {
+        assert!(exists<NonceHistory>(@aptos_framework), error::invalid_state(E_NONCE_HISTORY_DOES_NOT_EXIST));
+        let nonce_history = &mut NonceHistory[@aptos_framework];
+        let nonce_key = NonceKey {
+            sender_address,
+            nonce,
+        };
+        let index = sip_hash_from_value(&nonce_key) % NUM_BUCKETS;
+        let current_time = timestamp::now_seconds();
+        if (!nonce_history.nonce_table.contains(index)) {
+            nonce_history.nonce_table.add(
+                index,
+                empty_bucket()
+            );
+        };
+        let bucket = table::borrow_mut(&mut nonce_history.nonce_table, index);
+        // Garbage collect expired nonces in the bucket.
+        // If the last time a nonce was added to a map is over MAX_EXPIRATION_TIME ago, then the entire map has only expired nonces.
+        // So we can clear the map.
+        for (i in 0..2) {
+            if (bucket.nonce_maps[i].last_stored_time + MAX_EXPIRATION_TIME < current_time) {
+               bucket.nonce_maps[i].nonces.for_each_and_clear(|_, _| {});
+            };
+        };
+
+        // We call one of the two maps in the bucket as an "active" map.
+        // We check if the (address, nonce) pair exists in the both the maps in the bucket.
+        // But we only insert the (address, nonce) pair in the active map.
+        // Once every MAP_SWITCH_TIME seconds, we switch the active map.
+        // This means, after 60 seconds of using the current active map, all the entries in the other map
```

Comment:
the constant is 75 why it's 60 here?

Reply:
Updated the comment to use `MAX_EXPIRATION_TIME` instead of `60`.
If the current active map is used for `MAX_EXPIRATION_TIME` seconds, then that means we haven't added any nonces to the other inactive map for that long. All the nonces in the other inactive map must have been expired and can be garbage collected.

Reply:
I'm switching the "active" tag between maps once every 75 seconds (slightly more than the max expiration time), just to add some slack.